mkosi/__init__.py

```diff
@@ -7794,9 +7801,36 @@ def run_qemu_cmdline(config: MkosiConfig) -> Iterator[List[str]]:
         yield cmdline
 
 
+@contextlib.contextmanager
+def start_swtpm() -> Iterator[Optional[Path]]:
+
+    if not shutil.which("swtpm"):
+        MkosiPrinter.info("Couldn't find swtpm binary, not invoking qemu with TPM2 device.")
+        yield None
+        return
+
+    with tempfile.TemporaryDirectory() as swtpm_state:
+        swtpm_sock = Path(swtpm_state) / Path("sock")
+
+        cmd = ["swtpm",
+               "socket",
+               "--tpm2",
+               "--tpmstate", f"dir={swtpm_state}",
+               "--ctrl", f"type=unixio,path={swtpm_sock}",
+         ]
+
+        swtpm_proc = spawn(cmd)
+
+        yield swtpm_sock
+
+        swtpm_proc.wait()
+
+
 def run_qemu(config: MkosiConfig) -> None:
-    with run_qemu_cmdline(config) as cmdline:
-        run(cmdline, stdout=sys.stdout, stderr=sys.stderr)
+
+    with start_swtpm() as swtpm_socket:
```

Comment:
I would move this to within `run_qemu_cmdline()`, we probably need to rename it at some point since it does a bit more than that, but it's still the best fit for this stuff

Reply:
i am not sure we should bother with swtpm where we don't need it though? I'd just expose it where we need it, or where we expose a qemu machine to the user.

quite frankly, i don't understand why run_qemu_cmdline() is implemented as one of those context objects? i am by no means a python guru but i thought the point of context objects was that they can nicely auto-destroy stuff. but the cmdline?